library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,568 @@
+require "yast"
+
+require "abstract_method"
+
+module CWM
+  # Represent base for any widget used in CWM. It can be passed as "widget" argument. For more
+  # details about usage see {CWM.ShowAndRun}
+  #
+  # For using widgets design decision is to use subclassing. Reason is to have better separeated
+  # and easily reusable code. Opposite approach is to use instancing of existing classes, but
+  # especially with storing and initializing widgets it can be quite complex. Consider e.g.
+  # instancing of InputField like:
+  #
+  # ```
+  #   widget = InputField.new(
+  #     label: _("My label"),
+  #     help: _("blablbalba" \
+  #       "blablabla" \
+  #       "blablabla"
+  #     ),
+  #     init: Proc.new do |widget|
+  #       ...
+  #     end,
+  #     store: Proc.new do |widget, event|
+  #       ...
+  #     end,
+  #     validate: Proc.new do |widget, event|
+  #       ....
+  #     end
+  #   )
+  # ```
+  #
+  # For this example current subclassing approach looks like
+  # ```
+  #   class MyWidget < CWM::InputField
+  #     def label
+  #       _("My label")
+  #     end
+  #
+  #     def help
+  #       _("blablbalba" \
+  #         "blablabla" \
+  #         "blablabla"
+  #     end
+  #
+  #     def init
+  #       ...
+  #     end
+  #
+  #     def store
+  #       ...
+  #     end
+  #
+  #     def validate
+  #       ....
+  #     end
+  #   end
+  #  # and usage in dialog
+  #  widget = MyWidget.new
+  # ```
+  #
+  class AbstractWidget
+    include Yast::UIShortcuts
+    include Yast::I18n
+
+    # @return [String] id used for widget
+    attr_writer :widget_id
+    attr_writer :handle_all_events
+
+    # specify if widget handle all raised events or only its own
+    # By default only own values are handled
```

Comment:
> \# By default, {#handle} has no argument and it is called only for events of its own widget.
\# If `true`, {#handle}(event) is called for events of any widget.
\# @return [Boolean]